### PR TITLE
Fix WAL file removing on mirrors in archive_mode `always`

### PR DESCRIPTION
**Pull request by InnerLife0**
Far ago, we've backported (fa759d1bf70c020f884f05cf7ae15c6b9b47b4b1) archive_mode `always` to 6x. Since that, it was tested only manually.
Later, there was a bug [found](https://github.com/greenplum-db/gpdb/pull/14189) around WAL recycling depends on archive_mode. Upstream backported a fix (4a5a95b3a62dadd73af73b76f8fed9deeea842a6), but the condition for `XLogArchivingAlways()` was not backported, because upstream doesn't support it. As a result, WAL files on mirrors may be considered as removable and can be actually removed during restart point. Backporting of 78ea8b5daab9237fd42d7a8a836c1c451765499f solves the problem, but new tap-test was also added to prove it works and to show such cases.
_____
The patch reverts code changes made by a commit backported by upstream. However, upstream commit contains test files which we want to keep.
_____
My idea is to not squash commits during merge keeping revert and backported commits as different entities.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/65720

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117961

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117963

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1117956

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1121122

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1180398

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1180399

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67061

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1184710

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1184711

**Comment by whitehawk** [comment]
Please fix the description according to checklist from wiki

**Comment by RekGRpth** [comment]
> Please fix the description according to checklist from wiki

I think that this is not necessary, because the resulting patch will still consist of three different commits, each with its own description, and will go to the merge.

However, I suggest that in the second commit message you explicitly indicate the hash of which commit it was a backport.

**Comment by whitehawk** [comment]
> I think that this is not necessary, because the resulting patch will still consist of three different commits, each with its own description, and will go to the merge.

In this case, if you are going to integrate the commits below 'as is' with their own description:
https://github.com/arenadata/gpdb/commit/51bfdec5e581b2b51fe975bb025bfe2942478058
https://github.com/arenadata/gpdb/commit/5084340492b5a24de888631fefc1e7f204dc6f66
https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7

each description should comply with the checklist. In 84897d62654a100abdc22744815f1349afc5e7d7 the alignment is a bit strange - there are many empty spaces that can be filled.

File: src/test/recovery/t/139_archive_mode_always.pl [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+# The idea of the test is to check that not archived WAL files are not removed
+# by creation of checkpoint (on master) or restartpoint (on standby).
+
+# Initialize master node, doing archives
+my $node_master = get_new_node('master');
+$node_master->init(
+	has_archiving    => 1,
+	allows_streaming => 1);
+my $backup_name = 'my_backup';
+my $master_connstr = $node_master->connstr;
```
Comment (whitehawk) [inline src/test/recovery/t/139_archive_mode_always.pl]: I do not see where $master_connstr is used. What is the purpose of it?

File: src/test/recovery/t/139_archive_mode_always.pl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+use warnings;
+use PostgresNode;
+use TestLib;
+use Test::More tests => 2;
+use File::Copy;
+
+# The idea of the test is to check that not archived WAL files are not removed
+# by creation of checkpoint (on master) or restartpoint (on standby).
+
+# Initialize master node, doing archives
```
Comment (whitehawk) [inline src/test/recovery/t/139_archive_mode_always.pl:12]: Mixing continious and non-continious forms of verbs sounds a bit awkward to me. I would suggest to re-phrase, for ex. "Initialize master node, doing archives" -> "Initialize master node and make archives", or smth similar

File: src/test/recovery/t/139_archive_mode_always.pl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+create_restartpoint_on_ckpt_record_replay = 'on'
+});
+
+# Start it
+$node_master->start;
+
+# Take backup for standby
+$node_master->backup($backup_name);
+
+# Initialize standby node from backup, fetching WAL from archives
```
Comment (whitehawk) [inline src/test/recovery/t/139_archive_mode_always.pl:43]: The same as above - "Mixing continious and non-continious forms of verbs"

File: src/test/recovery/t/139_archive_mode_always.pl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+$node_master->safe_psql('postgres',
+	"CREATE TABLE t AS SELECT generate_series(1,1000) i;");
+
+# Find the next WAL segment to be archived
+my $walfile_to_be_kept = $node_master->safe_psql('postgres',
+	"SELECT pg_xlogfile_name(pg_current_xlog_location());");
+
+# Switch WAL and do checkpoint. Before version 11, Postgres needs files for
+# two points, so do checkpoint twice (see CreateCheckPoint() for more).
+# Checkpoint requested with CHECKPOINT_WAIT, so it's enough to wait for command
```
Comment (whitehawk) [inline src/test/recovery/t/139_archive_mode_always.pl]: "Checkpoint requested" -> "Checkpoint is requested" ? I suppose passive voice was meant

**Comment by InnerLife0** [comment]
@whitehawk Can you push all comments all at once?
Pushing them early makes no sense and looks like a spam.

File: src/test/recovery/t/139_archive_mode_always.pl (hunk 1 above)
Comment (InnerLife0) [reply src/test/recovery/t/139_archive_mode_always.pl]: True. This variable is redundant.

File: src/test/recovery/t/139_archive_mode_always.pl (hunk 2 above)
Comment (InnerLife0) [reply src/test/recovery/t/139_archive_mode_always.pl:12]: 1. Sorry, but I can't find any rule about it. If it's only a personal opinion, then I will ignore it, as none of us is native. If it's a rule, then please give a link.
2. The original comment was left by Alvaro Herrera in his [patch](https://github.com/postgres/postgres/commit/49148645f7f30f461b09618b5342b37425f33b22#diff-b57990e9ef44c06b531f9dd2279d95e1a344348d3298f81820a4eb4690fcb205R9) to PG. He, the author and other reviewers are probably not natives too, but all of them are OK with this.
3. (Deep) IMO, but we shouldn't care SO much about such things here or in any other patch (doesn't matter whether we made this changes or not).

File: src/test/recovery/t/139_archive_mode_always.pl (hunk 3 above)
Comment (InnerLife0) [reply src/test/recovery/t/139_archive_mode_always.pl:43]: The same as above.

File: src/test/recovery/t/139_archive_mode_always.pl (hunk 4 above)
Comment (InnerLife0) [reply src/test/recovery/t/139_archive_mode_always.pl]: Will fix this. Thanks.

**Comment by InnerLife0** [comment]
> each description should comply with the checklist. In [84897d6](https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7) the alignment is a bit strange - there are many empty spaces that can be filled.

What tool are you using to check commit message?

image

**Comment by whitehawk** [comment]
> What tool are you using to check commit message?

I've checked on web 
image

I see, it spoiled the view.

**Comment by RekGRpth** [comment]
> > each description should comply with the checklist. In [84897d6](https://github.com/arenadata/gpdb/commit/84897d62654a100abdc22744815f1349afc5e7d7) the alignment is a bit strange - there are many empty spaces that can be filled.
> 
> What tool are you using to check commit message?
> 
> image

Reduce commit hashes to the minimum value in the commit message.

**Comment by InnerLife0** [comment]
> Reduce commit hashes to the minimum value in the commit message.

Is it a rule? Some time ago I saw examples of same short hashes for different commits. Github made incorrect link for this.

**Comment by InnerLife0** [comment]
The force push was done as commit message of second commit was changed. Now it contains a link to original commit.
I believe we shouldn't divide a head of first commit message, as it's done by revert automatically.
The message of third commit was also changed closer to PR description.
Mistakes in test file were fixed too.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67222

**Comment by whitehawk** [comment]
I would suggest rephrasing the commit description to avoid speaking in the first person (e.g. "we've backported", "I decided", etc). 
But as it is not in the checklist , it is up to author for now.

**Review by whitehawk — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193313

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193314

**Review by RekGRpth — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67427

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204844

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204845

**Pull request merged**